:seedling: Lint hack/tools

wallchristopher opened the pull request:
**What this PR does / why we need it**:

Added the hack/tools/release directory to be linted like the rest of the repo. Attempted to do the hack/tools directory but ran into typecheck errors stating that the imports were not packages. Per golangci-lint's docs, [you can't skip or ignore these errors](https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors). Instead of modifying the import, I opted for just scanning the actual code.

Something to note, some of the keys within the [.golangci.yml](https://github.com/metal3-io/ip-address-manager/blob/main/.golangci.yaml) at not valid such as the [issues.skip-dirs](https://github.com/metal3-io/ip-address-manager/blob/main/.golangci.yaml#L94) and issues.skip-files. When playing around with the configuration they were a gotcha that got me.

Addressed issues identified by the linters.

**Tested via**:

local: within hack/tools/release dir `golangci-lint run --build-tags=tools --modules-download-mode=readonly`
act: repo root `act -j golangci`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #767

tuminoid commented:
/ok-to-test

tuminoid commented:
> **What this PR does / why we need it**:
> 
> Added the hack/tools/release directory to be linted like the rest of the repo. Attempted to do the hack/tools directory but ran into typecheck errors stating that the imports were not packages. Per golangci-lint's docs, [you can't skip or ignore these errors](https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors). Instead of modifying the import, I opted for just scanning the actual code.

/cc @kashifest @lentzi90 
WDYT?

> Something to note, some of the keys within the [.golangci.yml](https://github.com/metal3-io/ip-address-manager/blob/main/.golangci.yaml) at not valid such as the [issues.skip-dirs](https://github.com/metal3-io/ip-address-manager/blob/main/.golangci.yaml#L94) and issues.skip-files. When playing around with the configuration they were a gotcha that got me.

The config was [recently moved](https://github.com/metal3-io/ip-address-manager/pull/749) there as the old location was deprecated, and this should be the correct place for the configs. I don't see any complaints for golangci-lint, using "make lint". 

> Addressed issues identified by the linters.

Those look legit changes to me, and especially the unused/if-else problems alerts me of this linting issue in the first place.

> **Tested via**:
> 
> local: within hack/tools/release dir `golangci-lint run --build-tags=tools --modules-download-mode=readonly` act: repo root `act -j golangci`

"make lint"  should be linting this, but currently it does not. If I checkout the old notes.go from main which does not have any fixes, it will not catch any of them. This needs to be fixed.

tuminoid commented on .github/workflows/golangci-lint.yml:
```diff
@@ -17,6 +17,10 @@ jobs:
         working-directory:
         - ""
         - api
+        - hack/tools/release
```
This does not get linted using this config.

wallchristopher replied:
I'm not sure what you mean. The /.golangci-lint.yml config is used when golangci-lint is run.

If meaning when running `make lint`, I completely missed adding it. Thank you for bringing that up. Should be good picked up now.

tuminoid commented on .github/workflows/golangci-lint.yml:
```diff
@@ -17,6 +17,10 @@ jobs:
         - api
+        - hack/tools/release
+        include:
+          - working-directory: hack/tools/release
+            additional-args: --build-tags=tools --modules-download-mode=readonly
```
```suggestion
        - working-directory: hack/tools/release
          additional-args: --build-tags=tools --modules-download-mode=readonly
```
All YAML is indented in k8s style.

lentzi90 reviewed (commented):
Looks solid!
I think the commits should be squashed. It is not necessary to have one for each little fix. Please fix the comments by @tuminoid also, then I think this is good to merge

wallchristopher commented:
@tuminoid thank you for pointing out the `make lint`. Updated to now lint the tools directory.

> The config was https://github.com/metal3-io/ip-address-manager/pull/749 there as the old location was deprecated, and this should be the correct place for the configs. I don't see any complaints for golangci-lint, using "make lint"

It might just be my text-editor. Golangci-lint's docs don't mention skip-dirs / skip-files anymore. [Digging within the golangci-lint's code](https://github.com/golangci/golangci-lint/blob/master/pkg/config/run.go#L30) it seems that it still works but with a deprecation notice. Only wanted to bring it up.
![image](https://github.com/user-attachments/assets/2e40cf85-182a-46e7-89ca-37e91b707ee6)

@lentzi90 Appreciate it, squashed the commits into one.

tuminoid commented:
> It might just be my text-editor. Golangci-lint's docs don't mention skip-dirs / skip-files anymore. [Digging within the golangci-lint's code](https://github.com/golangci/golangci-lint/blob/master/pkg/config/run.go#L30) it seems that it still works but with a deprecation notice. Only wanted to bring it up. 

You're correct, it is deprecated. I wonder why it doesn't complain about those after they were once fixed in https://github.com/metal3-io/ip-address-manager/pull/749 and then it did complain about them.

 There are some other settings deprecated as well, so might as well fix them in a follow-up PR. And bump to latest version also. Previous bump was done minimally to just get support for Go 1.23.

tuminoid commented:
/override metal3-centos-e2e-integration-test-main metal3-ubuntu-e2e-integration-test-main

tuminoid reviewed (commented):
/approve

This is good improvement. Thank you for contributing @wallchristopher !

Rozzii reviewed (commented):
/lgtm

Rozzii commented:
Thanks @wallchristopher
